### PR TITLE
[Graphbolt] Add adding reverse edges helper

### PR DESCRIPTION
**Pull request by peizhou001**
## Description
Add a helper to find reverse edges for a set of given edges, it is always called before finding exclude edges.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: python/dgl/graphbolt/impl/sampled_subgraph_impl.py [hunk 1]
```diff
@@ -145,7 +174,20 @@ def exclude_edges(
...
+    exclude_mode : Union[str, callable], optional
+        Whether and how to exclude dependencies related to the provided edges
+        in the argument ``edges``. Possible values are
+        - If None, for excluding only provided ``edges``.
+        - If "reverse", for excluding not only the provided ``edges`` but aslo
+          their reverse edges. In the case of a homogeneous graph, reverse
+          edges refer to edges with inverted source and destination node IDs.
+          In heterogeneous graphs, this encompasses reversing both node IDs
+          and types, as specified by the ``reverse_types`` argument.
+        - If User-defined exclusion rule. It is a callable with edges in the
```
Comment (frozenbugs) [inline python/dgl/graphbolt/impl/sampled_subgraph_impl.py:186]: It is not necessary to allow callable, `exclude_edges(subgraph, edges, callable)` is equivalent to `exclude_edges(subgraph, callable(edges))`.

File: python/dgl/graphbolt/impl/sampled_subgraph_impl.py [hunk 2]
```diff
@@ -145,7 +174,20 @@ def exclude_edges(
...
         of edge types and the corresponding edges to exclude.
-
+    exclude_mode : Union[str, callable], optional
+        Whether and how to exclude dependencies related to the provided edges
+        in the argument ``edges``. Possible values are
+        - If None, for excluding only provided ``edges``.
+        - If "reverse", for excluding not only the provided ``edges`` but aslo
+          their reverse edges. In the case of a homogeneous graph, reverse
+          edges refer to edges with inverted source and destination node IDs.
+          In heterogeneous graphs, this encompasses reversing both node IDs
```
Comment (frozenbugs) [inline python/dgl/graphbolt/impl/sampled_subgraph_impl.py:184]: There are some corner cases we need to be crystal clear in doc string:
1) If in heterogeneous graphs, the edges (A:E:A) has the same node type for both src and dst, users still need to specify the reverse_etypes map (A:E:A) - (A:E:A), is that right?
2) if in heterogeneous graphs, for edges (A:E:B), users want to exclude all reverse edge (B:*:A) regardless edge type, users need to input all type in reverse_etypes (A:E:B) - (B:*:A), is that right?

File: python/dgl/graphbolt/impl/sampled_subgraph_impl.py [hunk 3]
```diff
@@ -179,6 +221,8 @@ def exclude_edges(
         "The sampled subgraph and the edges to exclude should be both "
         "homogeneous or both heterogeneous."
     )
+    # Get edges to exclude.
+    edges = _find_exclude_edges(edges, exclude_mode, reverse_etypes)
```
Comment (czkkkkkk) [inline python/dgl/graphbolt/impl/sampled_subgraph_impl.py:225]: The name `_find_exclude_edges` is confusing. It is more like expanding or transforming the edge set to exclude. Moreover, I think the logic of this function is independent from `exclude_edges`. What about we call it before `exclude_edges`?

File: python/dgl/graphbolt/impl/sampled_subgraph_impl.py (hunk 2 above)
Comment (peizhou001) [reply python/dgl/graphbolt/impl/sampled_subgraph_impl.py:184]: Totally correct, add this in the docstring for clearity.

File: python/dgl/graphbolt/utils/sample_utils.py [hunk 4]
```diff
@@ -8,6 +8,63 @@
...
+        Edges to exclude. If sampled subgraph is homogeneous, then `edges`
+        should be a pair of tensors representing the edges to exclude. If
+        sampled subgraph is heterogeneous, then `edges` should be a dictionary
+        of edge types and the corresponding edges to exclude.
+    reverse_etypes : Dict[str, str], optional
+        The mapping from the original edge types to their reverse edge types.
+
+    Returns
+    -------
+    SampledSubgraphImpl
```
Comment (frozenbugs) [inline python/dgl/graphbolt/utils/sample_utils.py:35]: Update the return type?

File: python/dgl/graphbolt/utils/sample_utils.py [hunk 5]
```diff
@@ -8,6 +8,63 @@
 from ..base import etype_str_to_tuple
 
 
+def add_reverse_edges(edges, reverse_etypes=None):
```
Comment (frozenbugs) [inline python/dgl/graphbolt/utils/sample_utils.py:11]: add types for parameter list?

File: python/dgl/graphbolt/utils/sample_utils.py [hunk 6]
```diff
@@ -8,6 +8,63 @@
...
+    used before `exclude_edges` function to help find targeting edges.
+    Note: The found reverse edges may not really exists in the original graph.
+    And repeat edges could be added becasue reverse edges may already exists in
+    the `edges`.
+
+    Parameters
+    ----------
+    edges : Union[Dict[str, Tuple[torch.Tensor, torch.Tensor]],
+                Tuple[torch.Tensor, torch.Tensor]]
+        Edges to exclude. If sampled subgraph is homogeneous, then `edges`
```
Comment (frozenbugs) [inline python/dgl/graphbolt/utils/sample_utils.py:26]: Does this limit to edges to exclude?

Comment (peizhou001) [reply python/dgl/graphbolt/utils/sample_utils.py:26]: yes, it does.

File: python/dgl/graphbolt/utils/sample_utils.py [hunk 7]
```diff
@@ -8,6 +8,72 @@
 from ..base import etype_str_to_tuple
 
 
+def find_reverse_edges(
```
Comment (czkkkkkk) [inline python/dgl/graphbolt/utils/sample_utils.py:11]: What about `add_reverse_edges`?

File: python/dgl/graphbolt/utils/sample_utils.py (hunk 6 above)
Comment (frozenbugs) [reply python/dgl/graphbolt/utils/sample_utils.py:32]: No, it does not :)

File: python/dgl/graphbolt/utils/sample_utils.py (hunk 7 above)
Comment (frozenbugs) [reply python/dgl/graphbolt/utils/sample_utils.py:11]: add_reverse_edges is slightly better than find_reverse_edges. It is pretty hard to find a name :(

Comment (peizhou001) [reply python/dgl/graphbolt/utils/sample_utils.py:11]: just changed back from `add_reverse_edges`... If both you think it's better, let me change back

File: python/dgl/graphbolt/utils/sample_utils.py [hunk 8]
```diff
@@ -8,6 +8,72 @@
...
+        Tuple[torch.Tensor, torch.Tensor],
+    ],
+    reverse_etypes: Dict[str, str] = None,
+):
+    r"""
+    This function finds the reverse edges of the given `edges` and returns the
+    composition of them. In a homogeneous graph, reverse edges have inverted
+    source and destination node IDs. While in a heterogeneous graph, reversing
+    also involves swapping node IDs and their types. This function is always
+    used before `exclude_edges` function to help find targeting edges.
```
Comment (frozenbugs) [inline python/dgl/graphbolt/utils/sample_utils.py:23]:
```suggestion
    also involves swapping node IDs and their types. This function could be used
    before `exclude_edges` function to help find targeting edges.
```

**Review by frozenbugs — APPROVED** [review]

**Pull request merged**